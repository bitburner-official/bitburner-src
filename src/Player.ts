import { sanitizeExploits } from "./Exploits/Exploit";

import { Reviver } from "./utils/JSONReviver";

import type { PlayerObject } from "./PersonObjects/Player/PlayerObject";

export let Player: PlayerObject;

export function setPlayer(playerObj: PlayerObject): void {
  Player = playerObj;
}

export function loadPlayer(saveString: string): PlayerObject {
  const player = JSON.parse(saveString, Reviver);
  player.money = parseFloat(player.money + "");
  player.exploits = sanitizeExploits(player.exploits);
  return player;
}
