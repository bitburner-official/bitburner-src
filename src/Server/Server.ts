// Class representing a single hackable Server
import { BaseServer } from "./BaseServer";

import { BitNodeMultipliers } from "../BitNode/BitNodeMultipliers";

import { createRandomString } from "../utils/helpers/createRandomString";
import { createRandomIp } from "../utils/IPAddress";
import { Generic_fromJSON, Generic_toJSON, IReviverValue, constructorsForReviver } from "../utils/JSONReviver";

export interface IConstructorParams {
  adminRights?: boolean;
  hackDifficulty?: number;
  hostname: string;
  ip?: string;
  isConnectedTo?: boolean;
  maxRam?: number;
  moneyAvailable?: number;
  numOpenPortsRequired?: number;
  organizationName?: string;
  purchasedByPlayer?: boolean;
  requiredHackingSkill?: number;
  serverGrowth?: number;
}

export class Server extends BaseServer {
  // Flag indicating whether this server has a backdoor installed by a player
  backdoorInstalled = false;

  // Initial server security level
  // (i.e. security level when the server was created)
  baseDifficulty = 1;

  // Server Security Level
  hackDifficulty = 1;

  // Minimum server security level that this server can be weakened to
  minDifficulty = 1;

  // How much money currently resides on the server and can be hacked
  moneyAvailable = 0;

  // Maximum amount of money that this server can hold
  moneyMax = 0;

  // Number of open ports required in order to gain admin/root access
  numOpenPortsRequired = 5;

  // How many ports are currently opened on the server
  openPortCount = 0;

  // Hacking level required to hack this server
  requiredHackingSkill = 1;

  // Parameter that affects how effectively this server's money can
  // be increased using the grow() Netscript function
  serverGrowth = 1;

  constructor(params: IConstructorParams = { hostname: "", ip: createRandomIp() }) {
    super(params);

    // "hacknet-node-X" hostnames are reserved for Hacknet Servers
    if (this.hostname.startsWith("hacknet-node-") || this.hostname.startsWith("hacknet-server-")) {
      this.hostname = createRandomString(10);
    }

    this.purchasedByPlayer = params.purchasedByPlayer != null ? params.purchasedByPlayer : false;

    //RAM, CPU speed and Scripts
    this.maxRam = params.maxRam != null ? params.maxRam : 0; //GB

    /* Hacking information (only valid for "foreign" aka non-purchased servers) */
    this.requiredHackingSkill = params.requiredHackingSkill != null ? params.requiredHackingSkill : 1;
    const baseMoney = params.moneyAvailable ?? 0;
    this.moneyAvailable = baseMoney * BitNodeMultipliers.ServerStartingMoney;
    this.moneyMax = 25 * baseMoney * BitNodeMultipliers.ServerMaxMoney;

    //Hack Difficulty is synonymous with server security. Base Difficulty = Starting difficulty
    const realDifficulty =
      params.hackDifficulty != null ? params.hackDifficulty * BitNodeMultipliers.ServerStartingSecurity : 1;
    this.hackDifficulty = Math.min(realDifficulty, 100);
    this.baseDifficulty = this.hackDifficulty;
    this.minDifficulty = Math.min(Math.max(1, Math.round(realDifficulty / 3)), 100);
    this.serverGrowth = params.serverGrowth != null ? params.serverGrowth : 1; //Integer from 0 to 100. Affects money increase from grow()

    //Port information, required for porthacking servers to get admin rights
    this.numOpenPortsRequired = params.numOpenPortsRequired != null ? params.numOpenPortsRequired : 5;
  }

  /** Ensures that the server's difficulty (server security) doesn't get too high */
  capDifficulty(): void {
    if (this.hackDifficulty < this.minDifficulty) {
      this.hackDifficulty = this.minDifficulty;
    }
    if (this.hackDifficulty < 1) {
      this.hackDifficulty = 1;
    }

    // Place some arbitrarily limit that realistically should never happen unless someone is
    // screwing around with the game
    if (this.hackDifficulty > 100) {
      this.hackDifficulty = 100;
    }
  }

  /**
   * Change this server's minimum security
   * @param n - Value by which to increase/decrease the server's minimum security
   * @param perc - Whether it should be changed by a percentage, or a flat value
   */
  changeMinimumSecurity(n: number, perc = false): void {
    if (perc) {
      this.minDifficulty *= n;
    } else {
      this.minDifficulty += n;
    }

    // Server security cannot go below 1
    this.minDifficulty = Math.max(1, this.minDifficulty);
  }

  /**
   * Change this server's maximum money
   * @param n - Value by which to change the server's maximum money
   */
  changeMaximumMoney(n: number): void {
    const softCap = 10e12;
    if (this.moneyMax > softCap) {
      const aboveCap = this.moneyMax - softCap;
      n = 1 + (n - 1) / Math.log(aboveCap) / Math.log(8);
    }

    this.moneyMax *= n;
  }

  /** Strengthens a server's security level (difficulty) by the specified amount */
  fortify(amt: number): void {
    this.hackDifficulty += amt;
    this.capDifficulty();
  }

  /** Lowers the server's security level (difficulty) by the specified amount) */
  weaken(amt: number): void {
    this.hackDifficulty -= amt * BitNodeMultipliers.ServerWeakenRate;
    this.capDifficulty();
  }

  /** Serialize the current object to a JSON save state */
  toJSON(): IReviverValue {
    return Generic_toJSON("Server", this);
  }

  // Initializes a Server Object from a JSON save state
  static fromJSON(value: IReviverValue): Server {
    return Generic_fromJSON(Server, value.data);
  }
}

constructorsForReviver.Server = Server;
