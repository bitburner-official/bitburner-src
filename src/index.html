<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <title><%= htmlWebpackPlugin.options.title %></title>
    <link rel="apple-touch-icon" sizes="180x180" href="dist/apple-touch-icon.png" />
    <link rel="icon" type="image/png" sizes="32x32" href="dist/favicon-32x32.png" />
    <link rel="icon" type="image/png" sizes="16x16" href="dist/favicon-16x16.png" />
    <link rel="manifest" href="dist/site.webmanifest" />
    <link rel="mask-icon" href="dist/safari-pinned-tab.svg" color="#000000" />
    <meta name="apple-mobile-web-app-title" content="Bitburner" />
    <meta name="application-name" content="Bitburner" />
    <meta name="msapplication-TileColor" content="#000000" />
    <meta name="msapplication-config" content="dist/browserconfig.xml" />
    <meta name="theme-color" content="#ffffff" />
    <style>
      html, body {
        margin: 0;
        padding: 0;
        height: 100%;
      }
      body {
        background-color: black;
      }
      * {
        -ms-overflow-style: none; /* for Internet Explorer, Edge */
        scrollbar-width: none; /* for Firefox */
      }
      .myGlyphMarginClass {
        background: red;
      }
      .myContentClass {
        background: lightblue;
      }
      *::-webkit-scrollbar {
        display: none; /* for Chrome, Safari, and Opera */
      }
    </style>
  </head>
  <body>
    <div id="root" style="display:flex" />
  </body>
</html>
