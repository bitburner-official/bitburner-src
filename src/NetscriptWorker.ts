/**
 * Functions for handling WorkerScripts, which are the underlying mechanism
 * that allows for scripts to run
 */
import { killWorkerScript } from "./Netscript/killWorkerScript";
import { ScriptDeath } from "./Netscript/ScriptDeath";
import { WorkerScript } from "./Netscript/WorkerScript";
import { workerScripts } from "./Netscript/WorkerScripts";
import { WorkerScriptStartStopEventEmitter } from "./Netscript/WorkerScriptStartStopEventEmitter";
import { generateNextPid } from "./Netscript/Pid";

import { CONSTANTS } from "./Constants";
import { Interpreter } from "./ThirdParty/JSInterpreter";
import { NetscriptFunctions } from "./NetscriptFunctions";
import { compile, Node } from "./NetscriptJSEvaluator";
import { Port, PortNumber } from "./NetscriptPort";
import { RunningScript } from "./Script/RunningScript";
import { scriptCalculateOfflineProduction } from "./Script/ScriptHelpers";
import { Script } from "./Script/Script";
import { GetAllServers } from "./Server/AllServers";
import { BaseServer } from "./Server/BaseServer";
import { Settings } from "./Settings/Settings";

import { generate } from "escodegen";

import { dialogBoxCreate } from "./ui/React/DialogBox";
import { arrayToString } from "./utils/helpers/arrayToString";
import { roundToTwo } from "./utils/helpers/roundToTwo";

import { parse } from "acorn";
import { simple as walksimple } from "acorn-walk";
import { areFilesEqual } from "./Terminal/DirectoryHelpers";
import { Terminal } from "./Terminal";
import { ScriptArg } from "@nsdefs";
import { handleUnknownError, CompleteRunOptions } from "./Netscript/NetscriptHelpers";

export const NetscriptPorts: Map<PortNumber, Port> = new Map();

export function prestigeWorkerScripts(): void {
  for (const ws of workerScripts.values()) {
    ws.env.stopFlag = true;
    killWorkerScript(ws);
  }

  NetscriptPorts.clear();

  WorkerScriptStartStopEventEmitter.emit();
  workerScripts.clear();
}

async function startNetscript2Script(workerScript: WorkerScript): Promise<void> {
  const scripts = workerScript.getServer().scripts;
  const script = workerScript.getScript();
  if (!script) throw "workerScript had no associated script. This is a bug.";
  const ns = workerScript.env.vars;
  if (!ns) throw `${script.filename} cannot be run because the NS object hasn't been constructed properly.`;

  const loadedModule = await compile(script, scripts);

  if (!loadedModule) throw `${script.filename} cannot be run because the script module won't load`;
  // TODO unplanned: Better error for "unexpected reserved word" when using await in non-async function?
  if (typeof loadedModule.main !== "function")
    throw `${script.filename} cannot be run because it does not have a main function.`;
  await loadedModule.main(ns);
}

async function startNetscript1Script(workerScript: WorkerScript): Promise<void> {
  const code = workerScript.code;
  let errorToThrow: unknown;

  //Process imports
  let codeWithImports, codeLineOffset;
  try {
    const importProcessingRes = processNetscript1Imports(code, workerScript);
    codeWithImports = importProcessingRes.code;
    codeLineOffset = importProcessingRes.lineOffset;
  } catch (e: unknown) {
    throw `Error processing Imports in ${workerScript.name}@${workerScript.hostname}:\n\n${e}`;
  }

  //TODO unplanned: Make NS1 wrapping type safe instead of using BasicObject.
  type BasicObject = Record<string, any>;
  const wrappedNS = NetscriptFunctions(workerScript);
  function wrapNS1Layer(int: Interpreter, intLayer: unknown, nsLayer = wrappedNS as BasicObject) {
    for (const [name, entry] of Object.entries(nsLayer)) {
      if (typeof entry === "function") {
        const wrapper = async (...args: unknown[]) => {
          try {
            // Sent a resolver function as an extra arg. See createAsyncFunction JSInterpreter.js:3209
            const callback = args.pop() as (value: unknown) => void;
            const result = await entry(...args.map((arg) => int.pseudoToNative(arg)));
            return callback(int.nativeToPseudo(result));
          } catch (e: unknown) {
            errorToThrow = e;
          }
        };
        int.setProperty(intLayer, name, int.createAsyncFunction(wrapper));
      } else if (Array.isArray(entry) || typeof entry !== "object") {
        // args, strings on enums, etc
        int.setProperty(intLayer, name, int.nativeToPseudo(entry));
      } else {
        // new object layer, e.g. bladeburner
        int.setProperty(intLayer, name, int.nativeToPseudo({}));
        wrapNS1Layer(int, (intLayer as BasicObject).properties[name], nsLayer[name]);
      }
    }
  }

  let interpreter: Interpreter;
  try {
    interpreter = new Interpreter(codeWithImports, wrapNS1Layer, codeLineOffset);
  } catch (e: unknown) {
    throw `Syntax ERROR in ${workerScript.name}@${workerScript.hostname}:\n\n${String(e)}`;
  }

  let more = true;
  while (more) {
    if (errorToThrow) throw errorToThrow;
    if (workerScript.env.stopFlag) return;
    for (let i = 0; more && i < 3; i++) more = interpreter.step();
    if (more) await new Promise((r) => setTimeout(r, Settings.CodeInstructionRunTime));
  }
}

/*  Since the JS Interpreter used for Netscript 1.0 only supports ES5, the keyword
    'import' throws an error. However, since we want to support import functionality
    we'll implement it ourselves by parsing the Nodes in the AST out.

    @param code - The script's code
    @returns {Object} {
        code: Newly-generated code with imported functions
        lineOffset: Net number of lines of code added/removed due to imported functions
                    Should typically be positive
    }
*/
function processNetscript1Imports(code: string, workerScript: WorkerScript): { code: string; lineOffset: number } {
  //allowReserved prevents 'import' from throwing error in ES5
  const ast: Node = parse(code, {
    ecmaVersion: 9,
    allowReserved: true,
    sourceType: "module",
  });

  const server = workerScript.getServer();
  if (server == null) {
    throw new Error("Failed to find underlying Server object for script");
  }

  function getScript(scriptName: string): Script | null {
    for (let i = 0; i < server.scripts.length; ++i) {
      if (server.scripts[i].filename === scriptName) {
        return server.scripts[i];
      }
    }
    return null;
  }

  let generatedCode = ""; // Generated Javascript Code
  let hasImports = false;

  // Walk over the tree and process ImportDeclaration nodes
  walksimple(ast, {
    ImportDeclaration: (node: Node) => {
      hasImports = true;
      let scriptName = node.source.value;
      if (scriptName.startsWith("./")) {
        scriptName = scriptName.slice(2);
      }
      const script = getScript(scriptName);
      if (script == null) {
        throw new Error("'Import' failed due to invalid script: " + scriptName);
      }
      const scriptAst = parse(script.code, {
        ecmaVersion: 9,
        allowReserved: true,
        sourceType: "module",
      });

      if (node.specifiers.length === 1 && node.specifiers[0].type === "ImportNamespaceSpecifier") {
        // import * as namespace from script
        const namespace = node.specifiers[0].local.name;
        const fnNames: string[] = []; //Names only
        const fnDeclarations: Node[] = []; //FunctionDeclaration Node objects
        walksimple(scriptAst, {
          FunctionDeclaration: (node: Node) => {
            fnNames.push(node.id.name);
            fnDeclarations.push(node);
          },
        });

        //Now we have to generate the code that would create the namespace
        generatedCode += `var ${namespace};\n(function (namespace) {\n`;

        //Add the function declarations
        fnDeclarations.forEach((fn: Node) => {
          generatedCode += generate(fn);
          generatedCode += "\n";
        });

        //Add functions to namespace
        fnNames.forEach((fnName) => {
          generatedCode += "namespace." + fnName + " = " + fnName;
          generatedCode += "\n";
        });

        //Finish
        generatedCode += `})(${namespace} || (" + namespace + " = {}));\n`;
      } else {
        //import {...} from script

        //Get array of all fns to import
        const fnsToImport: string[] = [];
        node.specifiers.forEach((e: Node) => {
          fnsToImport.push(e.local.name);
        });

        //Walk through script and get FunctionDeclaration code for all specified fns
        const fnDeclarations: Node[] = [];
        walksimple(scriptAst, {
          FunctionDeclaration: (node: Node) => {
            if (fnsToImport.includes(node.id.name)) {
              fnDeclarations.push(node);
            }
          },
        });

        //Convert FunctionDeclarations into code
        fnDeclarations.forEach((fn: Node) => {
          generatedCode += generate(fn);
          generatedCode += "\n";
        });
      }
    },
  });

  //If there are no imports, just return the original code
  if (!hasImports) {
    return { code: code, lineOffset: 0 };
  }

  //Remove ImportDeclarations from AST. These ImportDeclarations must be in top-level
  let linesRemoved = 0;
  if (ast.type !== "Program" || ast.body == null) {
    throw new Error("Code could not be properly parsed");
  }
  for (let i = ast.body.length - 1; i >= 0; --i) {
    if (ast.body[i].type === "ImportDeclaration") {
      ast.body.splice(i, 1);
      ++linesRemoved;
    }
  }

  //Calculated line offset
  const lineOffset = (generatedCode.match(/\n/g) || []).length - linesRemoved;

  //Convert the AST back into code
  code = generate(ast);

  //Add the imported code and re-generate in ES5 (JS Interpreter for NS1 only supports ES5);
  code = generatedCode + code;

  const res = {
    code: code,
    lineOffset: lineOffset,
  };
  return res;
}

/**
 * Used to start a RunningScript (by creating and starting its
 * corresponding WorkerScript), and add the RunningScript to the server on which
 * it is active
 */
export function startWorkerScript(runningScript: RunningScript, server: BaseServer, parent?: WorkerScript): number {
  if (createAndAddWorkerScript(runningScript, server, parent)) {
    // Push onto runningScripts.
    // This has to come after createAndAddWorkerScript() because that fn updates RAM usage
    server.runScript(runningScript);

    // Once the WorkerScript is constructed in createAndAddWorkerScript(), the RunningScript
    // object should have a PID assigned to it, so we return that
    return runningScript.pid;
  }

  return 0;
}

/**
 * Given a RunningScript object, constructs its corresponding WorkerScript,
 * adds it to the global 'workerScripts' pool, and begins executing it.
 * @param {RunningScript} runningScriptObj - Script that's being run
 * @param {Server} server - Server on which the script is to be run
 * returns {boolean} indicating whether or not the workerScript was successfully added
 */
function createAndAddWorkerScript(runningScriptObj: RunningScript, server: BaseServer, parent?: WorkerScript): boolean {
  const ramUsage = roundToTwo(runningScriptObj.ramUsage * runningScriptObj.threads);
  const ramAvailable = server.maxRam - server.ramUsed;
  // Check failure conditions before generating the workersScript and return false
  if (ramUsage > ramAvailable + 0.001) {
    dialogBoxCreate(
      `Not enough RAM to run script ${runningScriptObj.filename} with args ${arrayToString(runningScriptObj.args)}.\n` +
        `This can occur when you reload the game and the script's RAM usage has increased (either because of an update to the game or ` +
        `your changes to the script).\nThis can also occur if you have attempted to launch a script from a tail window with insufficient RAM. `,
    );
    return false;
  }

  // Get the pid
  const pid = generateNextPid();
  if (pid === -1) {
    dialogBoxCreate(
      `Failed to start script because could not find available PID. This is most ` +
        `because you have too many scripts running.`,
    );
    return false;
  }

  server.updateRamUsed(roundToTwo(server.ramUsed + ramUsage));

  // Create the WorkerScript. NOTE: WorkerScript ctor will set the underlying
  // RunningScript's PID as well
  const workerScript = new WorkerScript(runningScriptObj, pid, NetscriptFunctions);

  // Add the WorkerScript to the global pool
  workerScripts.set(pid, workerScript);
  WorkerScriptStartStopEventEmitter.emit();

  // Start the script's execution using the correct function for file type
  (workerScript.name.endsWith(".js") ? startNetscript2Script : startNetscript1Script)(workerScript)
    // Once the code finishes (either resolved or rejected, doesnt matter), set its
    // running status to false
    .then(function () {
      // On natural death, the earnings are transferred to the parent if it still exists.
      if (parent && !parent.env.stopFlag) {
        parent.scriptRef.onlineExpGained += runningScriptObj.onlineExpGained;
        parent.scriptRef.onlineMoneyMade += runningScriptObj.onlineMoneyMade;
      }
      killWorkerScript(workerScript);
      workerScript.log("", () => "Script finished running");
    })
    .catch(function (e) {
      handleUnknownError(e, workerScript);
      workerScript.log("", () => (e instanceof ScriptDeath ? "Script killed." : "Script crashed due to an error."));
      killWorkerScript(workerScript);
    });
  return true;
}

/** Updates the online running time stat of all running scripts */
export function updateOnlineScriptTimes(numCycles = 1): void {
  const time = (numCycles * CONSTANTS.MilliPerCycle) / 1000; //seconds
  for (const ws of workerScripts.values()) {
    ws.scriptRef.onlineRunningTime += time;
  }
}

/**
 * Called when the game is loaded. Loads all running scripts (from all servers)
 * into worker scripts so that they will start running
 */
export function loadAllRunningScripts(): void {
  const skipScriptLoad = window.location.href.toLowerCase().indexOf("?noscripts") !== -1;
  if (skipScriptLoad) {
    Terminal.warn("Skipped loading player scripts during startup");
    console.info("Skipping the load of any scripts during startup");
  }
  for (const server of GetAllServers()) {
    // Reset each server's RAM usage to 0
    server.ramUsed = 0;

    if (skipScriptLoad) {
      // Start game with no scripts
      server.runningScripts.length = 0;
      continue;
    }
    // Using backwards index iteration to avoid complications when removing elements during iteration.
    for (let i = server.runningScripts.length - 1; i >= 0; i--) {
      const runningScript = server.runningScripts[i];
      const success = createAndAddWorkerScript(runningScript, server);
      scriptCalculateOfflineProduction(runningScript);
      // Remove the RunningScript if the WorkerScript failed to start.
      if (!success) server.runningScripts.splice(i, 1);
    }
  }
}

/** Run a script from inside another script (run(), exec(), spawn(), etc.) */
export function runScriptFromScript(
  caller: string,
  host: BaseServer,
  scriptname: string,
  args: ScriptArg[],
  workerScript: WorkerScript,
  runOpts: CompleteRunOptions,
): number {
  /* Very inefficient, TODO change data structures so that finding script & checking if it's already running does
   * not require iterating through all scripts and comparing names/args. This is a big part of slowdown when
   * running a large number of scripts. */

  // Find the script, fail if it doesn't exist.
  const script = host.scripts.find((serverScript) => areFilesEqual(serverScript.filename, scriptname));
  if (!script) {
    workerScript.log(caller, () => `Could not find script '${scriptname}' on '${host.hostname}'`);
    return 0;
  }

  // Check if script is already running on server and fail if it is.
  if (host.getRunningScript(scriptname, args)) {
    workerScript.log(caller, () => `'${scriptname}' is already running on '${host.hostname}'`);
    return 0;
  }

  const singleRamUsage = runOpts.ramOverride ?? script.getRamUsage(host.scripts);
  if (!singleRamUsage) {
    workerScript.log(caller, () => `Ram usage could not be calculated for ${scriptname}`);
    return 0;
  }

  // Check if admin rights on host, fail if not.
  if (host.hasAdminRights == false) {
    workerScript.log(caller, () => `You do not have root access on '${host.hostname}'`);
    return 0;
  }

  // Calculate ram usage including thread count
  const ramUsage = singleRamUsage * runOpts.threads;

  // Check if there is enough ram to run the script, fail if not.
  const ramAvailable = host.maxRam - host.ramUsed;
  if (ramUsage > ramAvailable + 0.001) {
    workerScript.log(
      caller,
      () =>
        `Cannot run script '${scriptname}' (t=${runOpts.threads}) on '${host.hostname}' because there is not enough available RAM!`,
    );
    return 0;
  }
  // Able to run script
  workerScript.log(
    caller,
    () => `'${scriptname}' on '${host.hostname}' with ${runOpts.threads} threads and args: ${arrayToString(args)}.`,
  );
  const runningScriptObj = new RunningScript(script, singleRamUsage, args);
  runningScriptObj.threads = runOpts.threads;
  runningScriptObj.temporary = runOpts.temporary;

  return startWorkerScript(runningScriptObj, host, workerScript);
}
