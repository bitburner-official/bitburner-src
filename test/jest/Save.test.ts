import "../../src/Player";

import { loadAllServers, saveAllServers } from "../../src/Server/AllServers";
import { loadAllRunningScripts } from "../../src/NetscriptWorker";
import { Settings } from "../../src/Settings/Settings";

jest.useFakeTimers();

// Direct tests of loading and saving.
// Tests here should try to be comprehensive (cover as much stuff as possible)
// without requiring burdensome levels of maintenance when legitimate changes
// are made.

// Savegame generated from dev on 2023-03-12, mostly empty game with a few
// tweaks. A RunningScript was added in-game to test the one bit of
// non-trivial machinery involved in save/load.
//
// Most of the Servers have been removed to reduce space. Default values have
// been removed both for space, and to test that they are added correctly.
function loadStandardServers() {
  loadAllServers(String.raw`{
  "home": {
    "ctor": "Server",
    "data": {
      "hasAdminRights": true,
      "hostname": "home",
      "ip": "67.4.8.1",
      "isConnectedTo": true,
      "maxRam": 8,
      "messages": [
        "hackers-starting-handbook.lit"
      ],
      "organizationName": "Home PC",
      "programs": [
        "NUKE.exe"
      ],
      "ramUsed": 1.6,
      "runningScripts": [
        {
          "ctor": "RunningScript",
          "data": {
            "args": [],
            "filename": "script.js",
            "logs": [
              "I shouldn't even be saved, since I'm temporary"
            ],
            "logUpd": true,
            "offlineRunningTime": 0.01,
            "onlineRunningTime": 7.210000000000004,
            "pid": 3,
            "ramUsage": 1.6,
            "server": "home",
            "scriptKey": "script.js*[]",
            "temporary": true,
            "dependencies": [
              {
                "filename": "script.js",
                "url": "blob:http://localhost/302fe9e5-2ec3-4ed7-bb5a-4f8f4a85f46d",
                "moduleSequenceNumber": 2
              }
            ]
          }
        },
        {
          "ctor": "RunningScript",
          "data": {
            "args": [],
            "filename": "script.js",
            "logs": [
              "I'm a log line that should be pruned",
              "Another log line"
            ],
            "logUpd": true,
            "offlineRunningTime": 0.01,
            "onlineRunningTime": 7.210000000000004,
            "pid": 2,
            "ramUsage": 1.6,
            "server": "home",
            "scriptKey": "script.js*[]",
            "title": "Awesome Script",
            "dependencies": [
              {
                "filename": "script.js",
                "url": "blob:http://localhost/302fe9e5-2ec3-4ed7-bb5a-4f8f4a85f46d",
                "moduleSequenceNumber": 2
              }
            ]
          }
        }
      ],
      "scripts": {
        "ctor": "JSONMap",
        "data": [
          [
            "script.js",
            {
              "ctor": "Script",
              "data": {
                "code": "/** @param {NS} ns */\\nexport async function main(ns) {\\n  return ns.asleep(1000000);\\n}",
                "filename": "script.js",
                "module": {},
                "dependencies": [
                  {
                    "filename": "script.js",
                    "url": "blob:http://localhost/e0abfafd-2c73-42fc-9eea-288c03820c47",
                    "moduleSequenceNumber": 5
                  }
                ],
                "ramUsage": 1.6,
                "server": "home",
                "moduleSequenceNumber": 5,
                "ramUsageEntries": [
                  {
                    "type": "misc",
                    "name": "baseCost",
                    "cost": 1.6
                  }
                ]
              }
            }
          ]
        ]
      },
      "serversOnNetwork": [
        "n00dles"
      ],
      "purchasedByPlayer": true
    }
  },
  "n00dles": {
    "ctor": "Server",
    "data": {
      "hostname": "n00dles",
      "ip": "61.6.6.2",
      "maxRam": 4,
      "organizationName": "Noodle Bar",
      "serversOnNetwork": [
        "home"
      ],
      "moneyAvailable": 70000,
      "moneyMax": 1750000,
      "numOpenPortsRequired": 0,
      "serverGrowth": 3000
    }
  }
}`); // Fix confused highlighting `
  loadAllRunningScripts();
}

test("load/saveAllServers", () => {
  // Feed a JSON object through loadAllServers/saveAllServers.
  // The object is a pruned set of servers that was extracted from a real (dev) game.
  jest.setSystemTime(123456789000);
  loadStandardServers();

  // Re-stringify with indenting for nicer diffs
  const result = saveAllServers();
  expect(JSON.stringify(JSON.parse(result), null, 2)).toMatchSnapshot();
});

test("load/saveAllServers pruning RunningScripts", () => {
  // Feed a JSON object through loadAllServers/saveAllServers.
  // The object is a pruned set of servers that was extracted from a real (dev) game.

  loadStandardServers();

  // Re-stringify with indenting for nicer diffs
  Settings.ExcludeRunningScriptsFromSave = true;
  const result = saveAllServers();
  expect(JSON.stringify(JSON.parse(result), null, 2)).toMatchSnapshot();
});
