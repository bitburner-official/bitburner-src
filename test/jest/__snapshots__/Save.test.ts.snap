// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`load/saveAllServers 1`] = `
"{
  "home": {
    "ctor": "Server",
    "data": {
      "contracts": [],
      "cpuCores": 1,
      "ftpPortOpen": false,
      "hasAdminRights": true,
      "hostname": "home",
      "httpPortOpen": false,
      "ip": "67.4.8.1",
      "isConnectedTo": true,
      "maxRam": 8,
      "messages": [
        "hackers-starting-handbook.lit"
      ],
      "organizationName": "Home PC",
      "programs": [
        "NUKE.exe"
      ],
      "scripts": {
        "ctor": "JSONMap",
        "data": [
          [
            "script.js",
            {
              "ctor": "Script",
              "data": {
                "code": "/** @param {NS} ns */\\\\nexport async function main(ns) {\\\\n  return ns.asleep(1000000);\\\\n}",
                "filename": "script.js",
                "server": "home"
              }
            }
          ]
        ]
      },
      "serversOnNetwork": [
        "n00dles"
      ],
      "smtpPortOpen": false,
      "sqlPortOpen": false,
      "sshPortOpen": false,
      "textFiles": {
        "ctor": "JSONMap",
        "data": []
      },
      "purchasedByPlayer": true,
      "backdoorInstalled": false,
      "baseDifficulty": 1,
      "hackDifficulty": 1,
      "minDifficulty": 1,
      "moneyAvailable": 0,
      "moneyMax": 0,
      "numOpenPortsRequired": 5,
      "openPortCount": 0,
      "requiredHackingSkill": 1,
      "serverGrowth": 1,
      "runningScripts": [
        {
          "ctor": "RunningScript",
          "data": {
            "args": [],
            "dataMap": {},
            "filename": "script.js",
            "offlineExpGained": 0,
            "offlineMoneyMade": 0,
            "offlineRunningTime": 123456789.01,
            "onlineExpGained": 0,
            "onlineMoneyMade": 0,
            "onlineRunningTime": 7.210000000000004,
            "ramUsage": 1.6,
            "server": "home",
            "scriptKey": "script.js*[]",
            "title": "Awesome Script",
            "threads": 1,
            "temporary": false
          }
        }
      ]
    }
  },
  "n00dles": {
    "ctor": "Server",
    "data": {
      "contracts": [],
      "cpuCores": 1,
      "ftpPortOpen": false,
      "hasAdminRights": false,
      "hostname": "n00dles",
      "httpPortOpen": false,
      "ip": "61.6.6.2",
      "isConnectedTo": false,
      "maxRam": 4,
      "messages": [],
      "organizationName": "Noodle Bar",
      "programs": [],
      "scripts": {
        "ctor": "JSONMap",
        "data": []
      },
      "serversOnNetwork": [
        "home"
      ],
      "smtpPortOpen": false,
      "sqlPortOpen": false,
      "sshPortOpen": false,
      "textFiles": {
        "ctor": "JSONMap",
        "data": []
      },
      "purchasedByPlayer": false,
      "backdoorInstalled": false,
      "baseDifficulty": 1,
      "hackDifficulty": 1,
      "minDifficulty": 1,
      "moneyAvailable": 70000,
      "moneyMax": 1750000,
      "numOpenPortsRequired": 0,
      "openPortCount": 0,
      "requiredHackingSkill": 1,
      "serverGrowth": 3000,
      "runningScripts": []
    }
  }
}"
`;

exports[`load/saveAllServers pruning RunningScripts 1`] = `
"{
  "home": {
    "ctor": "Server",
    "data": {
      "contracts": [],
      "cpuCores": 1,
      "ftpPortOpen": false,
      "hasAdminRights": true,
      "hostname": "home",
      "httpPortOpen": false,
      "ip": "67.4.8.1",
      "isConnectedTo": true,
      "maxRam": 8,
      "messages": [
        "hackers-starting-handbook.lit"
      ],
      "organizationName": "Home PC",
      "programs": [
        "NUKE.exe"
      ],
      "scripts": {
        "ctor": "JSONMap",
        "data": [
          [
            "script.js",
            {
              "ctor": "Script",
              "data": {
                "code": "/** @param {NS} ns */\\\\nexport async function main(ns) {\\\\n  return ns.asleep(1000000);\\\\n}",
                "filename": "script.js",
                "server": "home"
              }
            }
          ]
        ]
      },
      "serversOnNetwork": [
        "n00dles"
      ],
      "smtpPortOpen": false,
      "sqlPortOpen": false,
      "sshPortOpen": false,
      "textFiles": {
        "ctor": "JSONMap",
        "data": []
      },
      "purchasedByPlayer": true,
      "backdoorInstalled": false,
      "baseDifficulty": 1,
      "hackDifficulty": 1,
      "minDifficulty": 1,
      "moneyAvailable": 0,
      "moneyMax": 0,
      "numOpenPortsRequired": 5,
      "openPortCount": 0,
      "requiredHackingSkill": 1,
      "serverGrowth": 1,
      "runningScripts": []
    }
  },
  "n00dles": {
    "ctor": "Server",
    "data": {
      "contracts": [],
      "cpuCores": 1,
      "ftpPortOpen": false,
      "hasAdminRights": false,
      "hostname": "n00dles",
      "httpPortOpen": false,
      "ip": "61.6.6.2",
      "isConnectedTo": false,
      "maxRam": 4,
      "messages": [],
      "organizationName": "Noodle Bar",
      "programs": [],
      "scripts": {
        "ctor": "JSONMap",
        "data": []
      },
      "serversOnNetwork": [
        "home"
      ],
      "smtpPortOpen": false,
      "sqlPortOpen": false,
      "sshPortOpen": false,
      "textFiles": {
        "ctor": "JSONMap",
        "data": []
      },
      "purchasedByPlayer": false,
      "backdoorInstalled": false,
      "baseDifficulty": 1,
      "hackDifficulty": 1,
      "minDifficulty": 1,
      "moneyAvailable": 70000,
      "moneyMax": 1750000,
      "numOpenPortsRequired": 0,
      "openPortCount": 0,
      "requiredHackingSkill": 1,
      "serverGrowth": 3000,
      "runningScripts": []
    }
  }
}"
`;
